Source/Core/VideoCommon/TextureInfo.cpp

```diff
@@ -0,0 +1,325 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "VideoCommon/TextureInfo.h"
+
+#include <xxhash.h>
+
+#include "Common/Align.h"
+#include "Core/HW/Memmap.h"
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/SamplerCommon.h"
+#include "VideoCommon/TextureDecoder.h"
+
+TextureInfo::TextureInfo(u32 stage)
+{
+  const FourTexUnits& tex = bpmem.tex[stage >> 2];
+  const u32 id = stage & 3;
+
+  m_texture_format = tex.texImage0[id].format;
+  m_tlut_format = tex.texTlut[id].tlut_format;
+
+  const bool is_palette_texture = IsColorIndexed(m_texture_format);
+  if (is_palette_texture)
+    m_palette_size = TexDecoder_GetPaletteSize(m_texture_format);
+
+  m_raw_width = tex.texImage0[id].width + 1;
+  m_raw_height = tex.texImage0[id].height + 1;
+
+  // TexelSizeInNibbles(format) * width * height / 16;
+  m_block_width = TexDecoder_GetBlockWidthInTexels(m_texture_format);
+  m_block_height = TexDecoder_GetBlockHeightInTexels(m_texture_format);
+
+  m_expanded_width = Common::AlignUp(m_raw_width, m_block_width);
+  m_expanded_height = Common::AlignUp(m_raw_height, m_block_height);
+
+  m_texture_size =
+      TexDecoder_GetTextureSizeInBytes(m_expanded_width, m_expanded_height, m_texture_format);
+
+  m_raw_address = (tex.texImage3[id].image_base /* & 0x1FFFFF*/) << 5;
+
+  m_from_tmem = tex.texImage1[id].cache_manually_managed != 0;
+  const u32 tmem_address_even = m_from_tmem ? tex.texImage1[id].tmem_even * TMEM_LINE_SIZE : 0;
+  const u32 tmem_address_odd = m_from_tmem ? tex.texImage2[id].tmem_odd * TMEM_LINE_SIZE : 0;
+  if (m_from_tmem)
+  {
+    m_data = &texMem[tmem_address_even];
+    m_tmem_odd = &texMem[tmem_address_odd];
+  }
+  else
+  {
+    m_data = Memory::GetPointer(m_raw_address);
+  }
+
+  const u32 tlutaddr = tex.texTlut[id].tmem_offset << 9;
+  m_tlut_address = &texMem[tlutaddr];
+
+  const bool has_mipmaps = SamplerCommon::AreBpTexMode0MipmapsEnabled(tex.texMode0[id]);
+  if (has_mipmaps)
+  {
+    const u32 raw_mip_count = (tex.texMode1[id].max_lod + 0xf) / 0x10;
+
+    // GPUs don't like when the specified mipmap count would require more than one 1x1-sized LOD in
+    // the mipmap chain
+    // e.g. 64x64 with 7 LODs would have the mipmap chain 64x64,32x32,16x16,8x8,4x4,2x2,1x1,0x0, so
+    // we limit the mipmap count to 6 there
+    const u32 mip_count =
+        std::min<u32>(IntLog2(std::max(m_raw_width, m_raw_height)) + 1, raw_mip_count + 1) - 1;
+
+    // load mips - TODO: Loading mipmaps from tmem is untested!
+    const u8* src_data = GetData() + GetTextureSize();
+    const u8* ptr_even = nullptr;
+    if (m_from_tmem)
+    {
+      ptr_even = &texMem[tmem_address_even + GetTextureSize()];
+    }
+    const u8* ptr_odd = m_tmem_odd;
+
+    for (u32 i = 0; i < mip_count; i++)
+    {
+      MipLevel mip_level(i + 1, *this, m_from_tmem, src_data, ptr_even, ptr_odd);
+      m_mip_levels.push_back(std::move(mip_level));
+    }
+  }
+}
+
+void TextureInfo::CalculateTextureName()
```

Comment:
It seems odd that this is a public function.  I'd design it by having the various name getters compute the data if needed, and use the `mutable` specifier on `m_name_cached` and the names.  At least as far as I can tell, `TextureInfo`'s actual data is immutable, so once the name has been computed once it will never need to be computed again (which is consistent with `m_name_cached` never being set to false again), so I think the actual design is fine.

Reply:
I personally hate the `mutable` keyword.  I just think it brings about confusion.  Hey this function is `const` but you have to look at the member variables to see if something actually changed!

That being said, I like your idea.  The only downside would be the getters would incur slight overhead by having multiple `if ( !m_name_cached ) CalculateTextureName()` calls.  I thought of making this function return a `tuple` or a `struct` instead and removing the getters.  Not sure if you'd prefer that or keeping the current design.

Reply:
I was thinking that instead of having `m_tlut_name`, `tlut_size` being nonzero would control whether or not the tlut hash (and the underscore before it) were included in the output.  But I don't think there's actually a good way of doing that.

Reply:
> Well, I was thinking I'd still cache it. There's a possibility the name can be used multiple times in one cycle

As I pointed on in my own review, you are burning a full 76 bytes for every `TextureInfo` object, just so you can save a few cpu cycles in a few rare code paths. I highly doubt the increased memory overhead is worth it.